src/shell-window-tracker.c

```diff
@@ -176,6 +184,71 @@ gboolean shell_window_tracker_is_speedwagon_window (MetaWindow *window)
   return g_strcmp0 (meta_window_get_role (window), SPEEDWAGON_ROLE) == 0;
 }
 
+static gboolean
+watchdog_timeout_reached_cb (gpointer user_data)
+{
+  ShellWindowTracker *self = user_data;
+
+  self->watchdog_id = 0;
+  if (self->coding_app)
+    self->coding_app = NULL;
+  return FALSE;
+}
+
+/**
+ * shell_window_tracker_track_coding_app_window:
+ * @tracker: An app monitor instance
+ * @app_window: A #MetaWindow
+ * Track a coding app.
+ *
+ */
+void shell_window_tracker_track_coding_app_window (ShellWindowTracker *tracker,
+                                                   MetaWindow *app_window)
+{
+  tracker->coding_app = app_window;
+  tracker->watchdog_id = g_timeout_add_seconds (WATCHDOG_TIMEOUT,
+                                                watchdog_timeout_reached_cb,
+                                                tracker);
+}
+
+/**
+ * shell_window_tracker_untrack_coding_app_window:
+ * @tracker: An app monitor instance
+ * @builder_window: A #MetaWindow
+ *
+ * Untrack a coding app.
+ *
+ */
+void shell_window_tracker_untrack_coding_app_window (ShellWindowTracker *tracker,
+                                                     MetaWindow *builder_window)
+{
+  MetaWindow *app_window;
+  app_window = g_hash_table_lookup (tracker->builder_to_app, builder_window);
```

Comment:
`g_hash_table_remove()` will already ignore things that are not in the table, so you don't need to lookup beforehand